commons/src/main/java/fr/free/nrw/commons/category/CategorizationFragment.java

```diff
@@ -87,113 +95,72 @@ public void writeToParcel(Parcel parcel, int flags) {
-            for(CategoryItem item : categoriesAdapter.getItems()) {
-                if(item.selected) {
-                    items.add(item);
-                    existingKeys.add(item.name);
-                }
-            }
-            for(String category : categories) {
-                if(!existingKeys.contains(category)) {
-                    items.add(new CategoryItem(category, false));
-                }
+                mergedItems.addAll(gpsItems);
             }
 
-            categoriesAdapter.setItems(items);
-            categoriesAdapter.notifyDataSetInvalidated();
-            categoriesSearchInProgress.setVisibility(View.GONE);
-            if (categories.size() == 0) {
-                if(TextUtils.isEmpty(filter)) {
-                    // If we found no recent cats, show the skip message!
-                    categoriesSkip.setVisibility(View.VISIBLE);
-                } else {
-                    categoriesNotFoundView.setText(getString(R.string.categories_not_found, filter));
-                    categoriesNotFoundView.setVisibility(View.VISIBLE);
-                }
-            } else {
-                categoriesList.smoothScrollToPosition(existingKeys.size());
-            }
+            mergedItems.addAll(items);
         }
+        catch (RemoteException e) {
+            throw new RuntimeException(e);
+        }
+        //Log.d(TAG, "Merged items: " + mergedItems.toString());
+        return mergedItems;
+    }
 
-        @Override
-        protected ArrayList<String> doInBackground(Void... voids) {
-            if(TextUtils.isEmpty(filter)) {
-                ArrayList<String> items = new ArrayList<String>();
-                ArrayList<String> mergedItems= new ArrayList<String>();
-
-                try {
-                    Cursor cursor = client.query(
-                            CategoryContentProvider.BASE_URI,
-                            Category.Table.ALL_FIELDS,
-                            null,
-                            new String[]{},
-                            Category.Table.COLUMN_LAST_USED + " DESC");
-                    // fixme add a limit on the original query instead of falling out of the loop?
-                    while (cursor.moveToNext() && cursor.getPosition() < SEARCH_CATS_LIMIT) {
-                        Category cat = Category.fromCursor(cursor);
-                        items.add(cat.getName());
-                    }
-
-                    if (MwVolleyApi.GpsCatExists.getGpsCatExists() == true){
-                        Log.d(TAG, "GPS cats found in CategorizationFragment.java" + MwVolleyApi.getGpsCat().toString());
-                        List<String> gpsItems = new ArrayList<String>(MwVolleyApi.getGpsCat());
-                        Log.d(TAG, "GPS items: " + gpsItems.toString());
 
-                        mergedItems.addAll(gpsItems);
-                    }
+    protected void setCatsAfterAsync(ArrayList<String> categories, String filter) {
```

Comment:
You might want to check that the Fragment is still added here. I think either Fragment.isAdded() or getActivity() != null.